### PR TITLE
Update dependencies to support Compose 1.5.0-beta01

### PR DESCRIPTION
**Pull request by malliaridis**
Update several library versions in `deps.versions.toml` to support Compose `1.5.0-beta01` (in favor to this [help request](https://blog.jetbrains.com/kotlin/2023/08/help-improve-compose-multiplatform-apps-performance/)).

Additional changes in `Pages` and `*.gradle.kts` files are necessary due to breaking changes in `ComposeCompilerKotlinSupportPlugin`, `rememberPagerState`, `HorizontalPager` and `VerticalPager`.

**IMPORTANT**: This commit introduces breaking changes for `Pages`.

_Note 1_: The Decompose version is not updated (I am not aware how the experimental branch is versioned / developed).
_Note 2_: Kotlin is not updated to 1.9.0 due to necessary changes in [gradle-setup-plugin](https://github.com/arkivanov/gradle-setup-plugin) (I couldn't manage to make it work with `mavenLocal()`).

### CODE REVIEW

**Comment by malliaridis** [comment]
@arkivanov Is this approach of PRs for the experimental branch correct? And what do you think about the breaking changes in `Pages`? We could abstract the `PagerScope` from Compose to avoid breaking changes in Decompose, but I am not sure if the scope may be useful for implementations down the path.

**Comment by arkivanov** [comment]
Thanks for the PR!

> Is this approach of PRs for the experimental branch correct?

In general, a stable version of Decompose must not depend on any non-stable libraries (e.g. Compose `1.5.0-beta01`). This includes the `compose-experimental` branch. I think we could include this change into `v2.1` branch instead, which is currently in alpha (the latest version is `2.1.0-alpha-06`).

> And what do you think about the breaking changes in `Pages`?

This should be OK, because the `Pages` API is marked with `ExperimentalDecomposeApi`. This would be just mentioned in the release notes.

File: deps.versions.toml [hunk 1]
```diff
@@ -1,25 +1,25 @@
 [versions]
 
 decompose = "2.0.1-compose-experimental"
-kotlin = "1.8.20"
+kotlin = "1.8.21"
```
Comment (arkivanov) [inline deps.versions.toml:4]: Is this required or just on-the-fly?

File: deps.versions.toml [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
 essenty = "1.1.0"
 parcelizeDarwin = "0.1.4"
 reaktive = "1.2.3"
 junit = "4.13.2"
-jetbrainsCompose = "1.4.0"
+jetbrainsCompose = "1.5.0-beta01"
 jetbrainsKotlinWrappers = "1.0.0-pre.478"
-jetbrainsKotlinxCoroutines = "1.6.4"
+jetbrainsKotlinxCoroutines = "1.7.1"
```
Comment (arkivanov) [inline deps.versions.toml:11]: Same here. If this is unrelated to the PR and just on-the-fly, it's better to do it separately, describing the benefits. Same for all versions below.

File: deps.versions.toml [hunk 3]
```diff
@@ -1,25 +1,25 @@
...
 jetbrainsBinaryCompatibilityValidator = "0.12.1"
-jetpackCompose = "1.4.0"
-jetpackComposeCompiler = "1.4.5"
-androidGradle = "7.2.2"
-androidMaterial = "1.6.1"
+jetpackCompose = "1.5.0-beta01"
+jetpackComposeCompiler = "1.5.1"
+androidGradle = "7.3.1"
```
Comment (arkivanov) [inline deps.versions.toml:15]: I'm trying to keep this version in sync with https://github.com/arkivanov/gradle-setup-plugin/blob/master/build.gradle.kts#L23. If possible, let's keep this unchanged, I will update AGP on my own later.

File: deps.versions.toml (hunk 1 above)
Comment (malliaridis) [reply deps.versions.toml:4]: I looked up the beta01 dependencies to check if this is actually required (I just did it on-the-fly before). It says:

> This version of Compose Multiplatform is based on the next Jetpack Compose libraries:
> [Compiler 1.5.0](https://developer.android.com/jetpack/androidx/releases/compose-compiler#1.5.0)
> [Runtime 1.5.0-beta03](https://developer.android.com/jetpack/androidx/releases/compose-runtime#1.5.0-beta03)
> [UI 1.5.0-beta03](https://developer.android.com/jetpack/androidx/releases/compose-ui#1.5.0-beta03)
> [Foundation 1.5.0-beta03](https://developer.android.com/jetpack/androidx/releases/compose-foundation#1.5.0-beta03)
> [Material 1.5.0-beta03](https://developer.android.com/jetpack/androidx/releases/compose-material#1.5.0-beta03)
> [Material3 1.1.1](https://developer.android.com/jetpack/androidx/releases/compose-material3#1.1.1)

Kotlin was updated to 1.9.0 in Compose Compiler 1.5.0 (according to the [docs](https://developer.android.com/jetpack/androidx/releases/compose-kotlin)), so I guess it should probably be upgraded all together?

_Note: Upgrading Kotlin to 1.9.0 would require updates in the `gradle-setup-plugin`_

File: deps.versions.toml (hunk 2 above)
Comment (malliaridis) [reply deps.versions.toml:11]: This is very likely not necessary and I will revert the changes. :)

File: deps.versions.toml (hunk 3 above)
Comment (malliaridis) [reply deps.versions.toml:15]: According to previous comment, the versions of `jetpackCompose` and `jetpackComposeCompiler` should be matched to the ones used in Compose MP (which is not the case here). I can update the versions, but as you mentioned, additional updates for keeping projects in sync would be necessary if updating the AGP.

I couldn't find any compatibility information about AGP, but it seems to work fine with 7.2.2. So I can revert this update too. :)

**Comment by malliaridis** [comment]
> I think we could include this change into v2.1 branch instead, which is currently in alpha (the latest version is 2.1.0-alpha-06).

@arkivanov Should I rebase the changes to `v2.1` (because they are based on the `experimental-compose` branch? Or do you already have any plans for merging `experimental-compose` into the rest of the project in the upcoming versions?
